shopinvader_address/models/res_partner.py

```diff
@@ -0,0 +1,163 @@
+# Copyright 2023 ACSONE SA/NV
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+
+from odoo import _, models
+from odoo.exceptions import MissingError, UserError
+
+
+class ResPartner(models.Model):
+
+    _inherit = "res.partner"
+
+    def _ensure_shopinvader_billing_address_not_used(self) -> None:
+        """
+        Check if Billing Address is used on confirmed sale order
+        """
+        self.ensure_one()
+        sale_order = (
+            self.env["sale.order"]
+            .sudo()
+            .search(
+                [
+                    ("partner_invoice_id", "=", self.id),
+                    ("state", "in", ("done", "sale")),
+                ],
+                limit=1,
+            )
+        )
+        if len(sale_order) > 0:
+            raise UserError(
+                _(
+                    "Can not update billing addresses(%(address_id)d)"
+                    "because it is already used on confirmed sale order",
+                    address_id=self.id,
+                )
```

Comment:
If I understood correctly the third idiom :  https://github.com/OCA/odoo-community.org/blob/master/website/Contribution/CONTRIBUTING.rst#33idioms and https://github.com/OCA/pylint-odoo/pull/304
it seems better to do 

```suggestion
                _(
                    "Can not update billing addresses(%(address_id)d)"
                    "because it is already used on confirmed sale order"
                )
                %
                {
                    address_id: self.id,
                }
```

Reply:
@FrancoMaxime Your suggestion is no more the right way to do it into Odoo (At least in 16). The translate method has been modified to take as kwargs value to put in placeholders.

Reply:
@lmignon oh, perfect. thanks for the clarification! 
@AnizR Sorry for this comment!